Richard - Added Staff Form, Staff Form test, and Staff Form story

richardfang888 opened the pull request:
In this PR, we add a React component for a data entry form for the course staff, complete with a story and test file. As a user, I can now go onto the issue of making the Staff page accessible from the Courses data table and have the form be visible there.

Closes #9

richardfang888 committed:
rf- added Form, Form test, and Form story for Staff component

richardfang888 committed:
rf- fix minor typo in StaffForm test

richardfang888 committed:
rf- fix import typo in StaffForm test

richardfang888 committed:
rf- fix impoactually fixed the import typo in StaffForm test

richardfang888 committed:
rf- fixed more typos in StaffForm Test file, changed StaffForm id field to correct naming convention

richardfang888 committed:
rf- changed form labels in StaffForm

richardfang888 committed:
rf- fixed a text field search change in StaffForm Test

richardfang888 committed:
rf- commented out unused variables

richardfang888 committed:
rf- fixed typos in StaffForm story

richardfang888 committed:
rf- fixing stryker mutation tests for StaffForm

richardfang888 committed:
rf- fixing stryker mutation tests for StaffForm part 2

richardfang888 committed:
rf- fixing stryker mutation tests for StaffForm part 3

pconrad commented on frontend/src/main/components/Staff/StaffForm.js:
```diff
@@ -0,0 +1,90 @@
+          </Form.Group>
+        </Col>
+        <Col>
+          <Form.Group className="mb-3">
+            <Form.Label htmlFor="githubId">github ID</Form.Label>
+            <Form.Control
+              data-testid="StaffForm-githubId"
+              id="githubId"
+              type="text"
+              isInvalid={Boolean(errors.githubId)}
+              {...register("githubId", { required: true })}
+            />
+            <Form.Control.Feedback type="invalid">
+              {errors.githubId && 'githubId is required.'}
+            </Form.Control.Feedback>
+          </Form.Group>
+        </Col>
+      </Row>
+      <Row>
+        <Col>
+          <Button
+            type="submit"
+            data-testid="StaffForm-submit"
+          >
+            {buttonLabel}
+          </Button>
+          <Button
+            variant="Secondary"
+            onClick={() => navigate(-1)}
+            data-testid="StaffForm-cancel"
+          >
+            Cancel
+          </Button>
+        </Col>
+      </Row>
+    </Form>
+  );
+}
+
+export default StaffForm;
```
I think the only thing we want on this form is literally GithubId and a submit button and literally nothing else.

The form will be embedded on a page where the course id is already there.   We'll use that information to send the add to the backend.

pconrad reviewed (changes requested):
* Please see my comments on the team slack channel about PR descriptions and update this one accordingly.
* Please also update the issue to which this links by filling in the description (currently totally blank) and adding acceptance crtiera.
* Please also address the other concerns in the CR below.

richardfang888 committed:
rf- removed courseId from Staff from as asked by TAs

richardfang888 committed:
rf- edited staff form tests to remove courseId properly

richardfang888 committed:
rf- updated staff fixture to not have courseId field

richardfang888 committed:
rf- reverted previous

richardfang888 committed:
rf- edited StaffForm test to match other componenent form tests

richardfang888 committed:
Merge branch 'main' into Richard--Staff-CRUD

richardfang888 committed:
rf- changed form formatting moving githubId to new row

richardfang888 committed:
Merge branch 'Richard--Staff-CRUD' of github.com:ucsb-cs156-s24/proj-organic-s24-5pm-2 into Richard--Staff-CRUD

richardfang888 committed:
rf- minor changes trying things out (cant figure out while tests failing)

richardfang888 committed:
rf- added await waitFor to ensure the DOM updates before checking the error message

richardfang888 committed:
rf- addeundo changes to .env.SAMPLE

pconrad commented on frontend/src/main/components/Staff/StaffForm.js:
```diff
@@ -0,0 +1,78 @@
+  // Stryker restore all
+  const navigate = useNavigate();
+
+  // edit to add regex if needed; change Form register pattern and feedback to match
+  // const courseId_regex = '';
+  // const githubId_regex = '';
```
It's not clear what the intentions of these comments are.  
Can we remove them?

richardfang888 committed:
rf- removed confusing comments

pconrad reviewed (changes requested):
Please review the comments on the team slack channel here:
https://ucsb-cs156-s24.slack.com/archives/C06SC553MLY/p1716870268080979

And see if this needs any reworking.

Adding staff should always be done in the context of a course.

Someone is a TA or an LA *for a specific course*. They are not an LA or TA "in general".

So the form for adding a staff member needs to have the information about the course right on the page, so that we know which course we are adding a TA or LA to.

I would suggest that the URL should contain the id of the course and the information about that course should be displayed at the top of the page.

I would suggest, actually not even HAVING a separate page for adding staff. Instead, it should go on the course show page.   And there is no need for edit.  Only add and delete.

richardfang888 committed:
rf- changed Staff Form so that the value of courseId is implicit in the form

richardfang888 committed:
rf- added forgotten import statement

richardfang888 commented:
> Please review the comments on the team slack channel here: https://ucsb-cs156-s24.slack.com/archives/C06SC553MLY/p1716870268080979
> 
> And see if this needs any reworking.
> 
> Adding staff should always be done in the context of a course.
> 
> Someone is a TA or an LA _for a specific course_. They are not an LA or TA "in general".
> 
> So the form for adding a staff member needs to have the information about the course right on the page, so that we know which course we are adding a TA or LA to.
> 
> I would suggest that the URL should contain the id of the course and the information about that course should be displayed at the top of the page.
> 
> I would suggest, actually not even HAVING a separate page for adding staff. Instead, it should go on the course show page. And there is no need for edit. Only add and delete.

I think I have addressed these changes in my most recent commit by making courseId implicit in the form and displayed alongside the id. I have adjusted the tests for the form as well.

abhishekambastha04 reviewed (approved):
LGTM

NevMan1 reviewed (approved):
LGTM

andrewpengucsb reviewed (approved):
LGTM
